src/middlewared/middlewared/plugins/pool.py

```diff
@@ -405,39 +405,45 @@ async def upgrade(self, oid):
     @private
     def transform_topology(self, x, options=None):
         """
-        Transform topology output from libzfs to add `device` and make `type` uppercase.
+        Take the top-level `groups` key from zfs.pool.query output and do 3 things:
+        1.  take the `type` key and upper-case it
+        2.  add `device` key that maps the gptid label to a partition (i.e. /dev/da1p1)
+        3.  add `disk` key that maps the gptid label to a disk (i.e /dev/da1)
         """
-        options = options or {}
-        if isinstance(x, dict):
-            if options.get('device_disk', True):
-                if 'label_to_dev_disk_cache' not in options:
-                    options['label_to_dev_disk_cache'] = self.middleware.call_sync('disk.label_to_dev_disk_cache')
-                path = x.get('path')
-                if path is not None:
-                    device = disk = None
-                    if path.startswith('/dev/'):
-                        args = [path[5:], False, options['label_to_dev_disk_cache']]
-                        device = self.middleware.call_sync('disk.label_to_dev', *args)
-                        disk = self.middleware.call_sync('disk.label_to_disk', *args)
-                    x['device'] = device
-                    x['disk'] = disk
-
-            if options.get('unavail_disk', True):
-                guid = x.get('guid')
-                if guid is not None:
-                    unavail_disk = None
-                    if x.get('status') != 'ONLINE':
-                        unavail_disk = self.middleware.call_sync('disk.disk_by_zfs_guid', guid)
-                    x['unavail_disk'] = unavail_disk
-
-            for key in x:
-                if key == 'type' and isinstance(x[key], str):
-                    x[key] = x[key].upper()
-                else:
-                    x[key] = self.transform_topology(x[key], dict(options, geom_scan=False))
-        elif isinstance(x, list):
-            for i, entry in enumerate(x):
-                x[i] = self.transform_topology(x[i], dict(options, geom_scan=False))
+        info = self.middleware.call_sync('disk.label_to_dev_and_disk')
```

Comment:
Can you please clarify what we are accomplishing apart from removing the recursion ? My concern is not having changes introduced unless really required to stable branch :)